### PR TITLE
Implement upserts

### PR DESCRIPTION
**Pull request by lilsweetcaligula**
Example usage (assuming the promisifying plugin is plugged in):
```
await Seneca
    .make(‘foo’)
    .data$({ x: 1, y: 2 })
    .save$({ upsert$: [‘x’] })
```

Questions:
- [ ] Where may I add documentation to explain the feature?
- [ ] Do we ignore the `upsert$` directive when the `#save$` method is being invoked on an existing entity? E.g.:
```
const existing_foo = await Seneca.make('foo').load$(foo_id);

Assert(existing_foo, 'existing_foo');

await existing_foo
    .data$({ x: 1, y: 2 })
    .save$({ upsert$: [‘x’] })
```

P.S. I believe the next step could be additional behavior testing of the features (such as, e.g., insertion, update, etc.) in order to probe for all possible edge cases and such.

### CODE REVIEW

File: test/mem.test.js [hunk 1]
```diff
@@ -269,6 +283,442 @@ describe('mem-store tests', function () {
...
+                  price: '3.40'
+                })
+
+                return fin()
+              })
+            })
+        })
+      })
+
+      describe('when some documents/records in the upsert$ directive match', () => {
```
Comment (rjrodger) [inline test/mem.test.js]: Upserts should only affect one record. We accept that this may simply be the first record found - but this is actually just following mongo: https://docs.mongodb.com/manual/reference/method/db.collection.updateOne/#behavior

File: test/mem.test.js [hunk 2]
```diff
@@ -269,6 +283,442 @@ describe('mem-store tests', function () {
       })
     })
   })
+
+  describe('upsert', () => {
```
Comment (rjrodger) [inline test/mem.test.js:288]: we need a test case where the upsert field is another unique field - e.g. users with id and username, with the upsert on username - this is the "happy path" test - demonstrate that the existing user is updated rather than a new one created - make this the first test

File: test/mem.test.js [hunk 3]
```diff
@@ -269,6 +283,442 @@ describe('mem-store tests', function () {
...
+    describe('when creating a new document/record', () => {
```
Comment (rjrodger) [inline test/mem.test.js]: a critical test for this functionality is using it in seneca-vote - I have created an issue:
https://github.com/lilsweetcaligula/seneca-vote/issues/38

File: test/mem.test.js [hunk 4]
```diff
@@ -269,6 +283,442 @@ describe('mem-store tests', function () {
...
+
+                app.make('product').list$({}, (err, products) => {
+                  if (err) {
+                    return fin(err)
+                  }
+
+                  expect(products.length).to.equal(2)
+
+                  expect(products[0]).to.contain({
+                    // NOTE: WARNING: Seneca seems to be stripping out fields
```
Comment (rjrodger) [inline test/mem.test.js]: This is actually by design. Use null instead as that represents the DB data model more accurately.

File: test/mem.test.js [hunk 5]
```diff
@@ -269,6 +283,442 @@ describe('mem-store tests', function () {
...
+      describe('when some of the fields in the upsert$ directive do not exist in the attributes passed via data$', () => {
+        const app = makeSenecaForTest()
+
+        beforeEach(fin => {
+          app.make('product')
+            .data$({ label: 'a toothbrush', price: '3.40' })
+            .save$(fin)
+        })
+
+        it('ignores the missing fields and updates the matching existing record', fin => {
```
Comment (rjrodger) [inline test/mem.test.js]: This is incorrect - all upsert fields must match exactly, otherwise create a new record

File: mem-store.ts [hunk 6]
```diff
@@ -106,18 +107,129 @@ function mem_store(options: any) {
...
+              const qent = msg.ent;
+              const public_entdata = msg.ent.data$(false);
+
+              const q = upsert_on.reduce((acc: any, upsert_on_field: string) => {
+                acc[upsert_on_field] = public_entdata[upsert_on_field];
+                return acc;
+              }, {});
+
+
+              return listents(seneca, entmap, qent, q, function (err: Error | null, docs_to_update: any[]) {
```
Comment (rjrodger) [inline mem-store.ts]: See note in test suite - this is not necessary as upsert should only affect one record at most. Since the upsert only supports exact matches, you can use Array.find on the  internal entity store directly, rather than listents

File: mem-store.ts [hunk 7]
```diff
@@ -437,3 +558,9 @@ function listents(seneca: any, entmap: any, qent: any, q: any, done: any) {
   // Return the resulting list to the caller.
   done.call(seneca, null, list)
 }
+
+function isPrivateProp(prop: any) : boolean {
```
Comment (rjrodger) [inline mem-store.ts]: this level of abstraction is too low - instead create a utility function in common.js called `clean` (for consistency with other parts of Seneca), that takes either a string[] or {[string]: any} and removes any elements or keys containg '$', returning a "clean" value

File: test/mem.test.js [hunk 8]
```diff
@@ -269,6 +283,442 @@ describe('mem-store tests', function () {
       })
     })
   })
+
+  describe('upsert', () => {
+    describe('when creating a new document/record', () => {
+      describe('when no documents/records in the upsert$ directive match', () => {
```
Comment (rjrodger) [inline test/mem.test.js]: create a more systematic ordering to the test cases to allow us to "eyeball" completeness - e.g.

* target does not exist
  * 0,1,2 upsert$ fields
    * new entity created
    * old entities not affected
* target does exist
  * 0,1,2 upsert$ fields
    * existing entity updated
    * no new entities
    * other entities not affected
* edge cases
  * null values, etc

File: mem-store.ts (hunk 6 above)
Comment (lilsweetcaligula) [reply mem-store.ts]: @rjrodger 
The implementation of the `load` method relies on the `listents` helper. Would we prefer the suggested simpler approach - with the `Array#find` suggestion, or a more "idiomatic" one in terms of the code already present in the module - with the usage of the `listents` helper?

File: test/mem.test.js (hunk 5 above)
Comment (lilsweetcaligula) [reply test/mem.test.js]: @rjrodger Do I understand the request correctly - do you mean, if a field is present in the `upsert$` array passed to the `#save$` method, but not inside the object passed to the `#data$` method, - then we should create a new record?

E.g.
```javascript
app.make('product')
  .data$({ label: 'a toothbrush' }) // The `coolness_factor` is missing here.
  .save$({ upsert$: ['label', 'coolness_factor'] }, cb) // The `coolness_factor` is specified inside the `upsert$` here.
```

File: mem-store.ts (hunk 7 above)
Comment (lilsweetcaligula) [reply mem-store.ts]: There already seems to be the implementation of `clean` in the `seneca.js` repo:
https://github.com/senecajs/seneca/blob/master/lib/common.js#L251

I was wondering if we should create a package of utilities and helpers for the Seneca eco-system that would basically be a collection of utilities like `clean`, etc.?
https://github.com/senecajs/seneca/blob/master/lib/common.js

File: test/mem.test.js (hunk 5 above)
Comment (rjrodger) [reply test/mem.test.js]: It depends what's in the db:
```
db=[{id:1,x:1,y:2,z:3}]  // note: resets for each below

save {x:1,y:22} upsert$:x => db=[{id:1,x:1,y:22,z:3}]   // update as upsert fields match
save {x:11,y:22} upsert$:x => db=[{id:1,x:1,y:22,z:3},{id:2,x:11,y:22}]   // insert as upsert fields do not match

save {x:1,y:2,z:33} upsert$:x,y => db=[{id:1,x:1,y:2,z:33}]   // update as upsert fields match
save {x:11,y:2,z:33} upsert$:x,y => db=[{id:1,x:1,y:2,z:33},{id:2,x:11,y:2,z:33}]   // insert as upsert fields do not match

// undef fields
save {x:1,z:33} upsert$:x,y => db=[{id:1,x:1,y:2,z:33},{id:2,x:1,z:33}] // insert as upsert fields do not match (undef != 2)
```

Update is strict - all upsert fields must match exactly - since undefineds should never match, you can short circuit to insert if any of the upsert fields are undef in the entity to save.

Another way to think about this is that if an ent has no id, insert is the default, and strongly preferred, and update has to  really work hard to happen

File: mem-store.ts (hunk 6 above)
Comment (rjrodger) [reply mem-store.ts]: load uses listents because it has to support complex queries (gt, lt, etc are possible), but upsert is much simpler - it must be an *exact* match, so you use find directly - this saves a callback level

File: mem-store.ts (hunk 7 above)
Comment (rjrodger) [reply mem-store.ts]: On the utils - Seneca.util serves that purpose in the API.

The Seneca clean function assumes an object, and treats arrays like objects. The clean function we need here is different - it cleans array values.

You could imagine extending Seneca.util.clean - BUT - at the framework level we must be ultra conservative. Each feature in a framework is very expensive (feature complexity grows as n^2 as feature interactions are a case of https://en.wikipedia.org/wiki/Metcalfe%27s_law), will last years, and may have unforeseen consequences. Our general strategy is thus: implement locally, observe, and consider over time if a feature has "earned" a place  in the framework.

Seneca.util already contains stuff that should not be there - eg. https://github.com/senecajs/seneca/blob/master/lib/common.js#L289 - why? why? now we must live with it (deprecate in Seneca 4, remove in Seneca 5) - frameworks!

File: package.json [hunk 9]
```diff
@@ -43,12 +43,13 @@
   "devDependencies": {
     "@hapi/code": "^8.0.2",
     "@hapi/lab": "^24.1.0",
+    "async": "3.2.0",
+    "lab-transform-typescript": "^3.0.1",
```
Comment (lilsweetcaligula) [inline package.json:47]: Q: Should we freeze the version?

File: mem-store.ts [hunk 10]
```diff
@@ -106,18 +108,97 @@ function mem_store(options: any) {
...
+          prev = entmap[base][name][mement.id] = mement
+
+          seneca.log.debug(function () {
+            return [
+              'save/' + (isNewEntityBeingCreated(msg) ? 'insert' : 'update'),
+              ent.canon$({ string: 1 }),
+              mement,
+              desc,
+            ]
+          })
```
Comment (lilsweetcaligula) [inline mem-store.ts:134]: Q: Should we log upserts in a similar manner as well? E.g. upon match:
```javascript
seneca.log.debug(() => [
  'save/update',
  ent.canon$({ string: 1 }),
  mement,
  desc
])
```

File: mem-store.ts [hunk 11]
```diff
@@ -106,18 +108,97 @@ function mem_store(options: any) {
...
+          //
+          const query_for_save = msg.q
+
+
+          if (isNewEntityBeingCreated(msg) && Array.isArray(query_for_save.upsert$)) {
+            const upsert_on = Common.clean(query_for_save.upsert$)
+
+
+            if (upsert_on.length > 0) {
+              if (!(base in entmap)) {
```
Comment (lilsweetcaligula) [inline mem-store.ts:158]: While technically this check is unnecessary due to: https://github.com/senecajs/seneca-mem-store/blob/ed2dec365ef55e6b589de7a80f69ff386a212007/mem-store.ts#L89

-I think it is nonetheless important to include it here. The reason is that in this manner `upsertIfRequested` is decoupled from the knowledge about the inner workings of the `do_save` function.

The only thing `upsertIfRequested` is concerned about, is that if there is no store for the requested entity, then there can be no match - and as a result - no update.

File: mem-store.ts [hunk 12]
```diff
@@ -106,18 +108,97 @@ function mem_store(options: any) {
...
+                  return reply(null, {
+                    did_upsert: true,
+                    out
+                  })
+              })
+            } else {
+              return reply(null, { did_upsert: false, out: null })
+            }
+          } else {
+            return reply(null, { did_upsert: false, out: null })
```
Comment (lilsweetcaligula) [inline mem-store.ts:199]: Note to self:
TODO:

rename `did_upsert` to `did_update`.

File: .gitignore [hunk 13]
```diff
@@ -17,4 +17,6 @@ node_modules
 .DS_Store
 coverage.html
 
-.history
\ No newline at end of file
+.history
+
+package-lock.json
```
Comment (lilsweetcaligula) [inline .gitignore:22]: Source: https://dev.to/gajus/stop-using-package-lock-json-or-yarn-lock-3ddi

File: package-lock.json [hunk 14]
```diff
@@ -1,2035 +0,0 @@
-{
```
Comment (lilsweetcaligula) [inline package-lock.json:1]: Source: https://dev.to/gajus/stop-using-package-lock-json-or-yarn-lock-3ddi

File: test/mem.test.js (hunk 1 above)
Comment (lilsweetcaligula) [reply test/mem.test.js]: Addressed:
https://github.com/senecajs/seneca-mem-store/blob/c9a4126230d2b3fb25adc6c51ab5ce5c13438914/mem-store.ts#L172
https://github.com/senecajs/seneca-mem-store/blob/c9a4126230d2b3fb25adc6c51ab5ce5c13438914/test/mem.test.js#L403

File: test/mem.test.js (hunk 2 above)
Comment (lilsweetcaligula) [reply test/mem.test.js:288]: Addressed:
https://github.com/senecajs/seneca-mem-store/blob/c9a4126230d2b3fb25adc6c51ab5ce5c13438914/test/mem.test.js#L308

File: test/mem.test.js (hunk 3 above)
Comment (lilsweetcaligula) [reply test/mem.test.js]: Added a test to verify that no race conditions occur:
https://github.com/senecajs/seneca-mem-store/blob/c9a4126230d2b3fb25adc6c51ab5ce5c13438914/test/mem.test.js#L626

File: test/mem.test.js (hunk 4 above)
Comment (lilsweetcaligula) [reply test/mem.test.js]: Addressed:
https://github.com/senecajs/seneca-mem-store/blob/c9a4126230d2b3fb25adc6c51ab5ce5c13438914/test/mem.test.js#L626

File: test/mem.test.js (hunk 5 above)
Comment (lilsweetcaligula) [reply test/mem.test.js]: Addressed:
https://github.com/senecajs/seneca-mem-store/blob/c9a4126230d2b3fb25adc6c51ab5ce5c13438914/test/mem.test.js#L626

File: mem-store.ts (hunk 6 above)
Comment (lilsweetcaligula) [reply mem-store.ts]: Addressed:
https://github.com/senecajs/seneca-mem-store/blob/c9a4126230d2b3fb25adc6c51ab5ce5c13438914/mem-store.ts#L172

File: mem-store.ts (hunk 7 above)
Comment (lilsweetcaligula) [reply mem-store.ts]: Addressed:
https://github.com/senecajs/seneca-mem-store/blob/c9a4126230d2b3fb25adc6c51ab5ce5c13438914/mem-store.ts#L154
https://github.com/senecajs/seneca-mem-store/blob/c9a4126230d2b3fb25adc6c51ab5ce5c13438914/lib/common.ts#L3

I was originally thinking to include tests for the `clean` helper, - to test for the scenario when an object of type `{ [string]: any }` is passed to it, however that would require that tests be setup for TypeScript, which can be an issue of its own.

Additionally, the current use case wherein an array of strings (i.e. `string[]`) is passed to `clean`, is already covered by the `upsert` tests in `test/mem.test.js`.

File: test/mem.test.js (hunk 8 above)
Comment (lilsweetcaligula) [reply test/mem.test.js]: Addressed, but please do take a look at `test/mem.test.js` to make sure I understood the request correctly.

**Comment by rjrodger** [comment]
final task to accept this: validate upserts in seneca-vote - show locking is not needed

File: package.json (hunk 9 above)
Comment (rjrodger) [reply package.json:47]: yes, and validate with https://github.com/rjrodger/seneca-plugin-validator - work for later
